From 2d0407186bf1ec4b3ce4ac54194fc4dffe18adb0 Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Wed, 22 Jan 2025 16:32:10 +0800
Subject: [PATCH 90/99] drm: Support HW cursor for non-GL renderers

Support hardware cursor on DRM backend without GL/Vulkan renderer by
using DRM dumb buffer exclusively.

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 libweston/backend-drm/drm-gbm.c       | 131 --------------------------
 libweston/backend-drm/drm-internal.h  |  19 +---
 libweston/backend-drm/drm.c           |  73 ++++++++++----
 libweston/backend-drm/fb.c            |  80 +++++++++-------
 libweston/backend-drm/kms.c           |   4 +-
 libweston/backend-drm/state-propose.c |  21 +----
 6 files changed, 105 insertions(+), 223 deletions(-)

diff --git a/libweston/backend-drm/drm-gbm.c b/libweston/backend-drm/drm-gbm.c
index da2198241..d7084856d 100644
--- a/libweston/backend-drm/drm-gbm.c
+++ b/libweston/backend-drm/drm-gbm.c
@@ -141,130 +141,6 @@ init_vulkan(struct drm_backend *b)
 	return 0;
 }
 
-static void drm_output_fini_cursor_egl(struct drm_output *output)
-{
-	unsigned int i;
-
-	for (i = 0; i < ARRAY_LENGTH(output->gbm_cursor_fb); i++) {
-		/* This cursor does not have a GBM device */
-		if (output->gbm_cursor_fb[i] && !output->gbm_cursor_fb[i]->bo)
-			output->gbm_cursor_fb[i]->type = BUFFER_PIXMAN_DUMB;
-		drm_fb_unref(output->gbm_cursor_fb[i]);
-		output->gbm_cursor_fb[i] = NULL;
-	}
-}
-
-static void drm_output_fini_cursor_vulkan(struct drm_output *output)
-{
-	unsigned int i;
-
-	for (i = 0; i < ARRAY_LENGTH(output->gbm_cursor_fb); i++) {
-		/* This cursor does not have a GBM device */
-		if (output->gbm_cursor_fb[i] && !output->gbm_cursor_fb[i]->bo)
-			output->gbm_cursor_fb[i]->type = BUFFER_PIXMAN_DUMB;
-		drm_fb_unref(output->gbm_cursor_fb[i]);
-		output->gbm_cursor_fb[i] = NULL;
-	}
-}
-
-static int
-drm_output_init_cursor_egl(struct drm_output *output, struct drm_backend *b)
-{
-	struct drm_device *device = output->device;
-	unsigned int i;
-
-	/* No point creating cursors if we don't have a plane for them. */
-	if (!output->cursor_handle)
-		return 0;
-
-	for (i = 0; i < ARRAY_LENGTH(output->gbm_cursor_fb); i++) {
-		struct gbm_bo *bo;
-
-		if (gbm_device_get_fd(b->gbm) != output->device->kms_device->fd) {
-			output->gbm_cursor_fb[i] =
-				drm_fb_create_dumb(output->device,
-						   device->cursor_width,
-						   device->cursor_height,
-						   DRM_FORMAT_ARGB8888);
-			/* Override buffer type, since we know it is a cursor */
-			output->gbm_cursor_fb[i]->type = BUFFER_CURSOR;
-			output->gbm_cursor_handle[i] =
-				output->gbm_cursor_fb[i]->handles[0];
-		} else {
-			bo = gbm_bo_create(b->gbm, device->cursor_width, device->cursor_height,
-					   GBM_FORMAT_ARGB8888,
-					   GBM_BO_USE_CURSOR | GBM_BO_USE_WRITE);
-			if (!bo)
-				goto err;
-
-			output->gbm_cursor_fb[i] =
-				drm_fb_get_from_bo(bo, device, false, BUFFER_CURSOR);
-			if (!output->gbm_cursor_fb[i]) {
-				gbm_bo_destroy(bo);
-				goto err;
-			}
-			output->gbm_cursor_handle[i] = gbm_bo_get_handle(bo).s32;
-		}
-	}
-
-	return 0;
-
-err:
-	weston_log("cursor buffers unavailable, using gl cursors\n");
-	device->cursors_are_broken = true;
-	drm_output_fini_cursor_egl(output);
-	return -1;
-}
-
-static int
-drm_output_init_cursor_vulkan(struct drm_output *output, struct drm_backend *b)
-{
-	struct drm_device *device = output->device;
-	unsigned int i;
-
-	/* No point creating cursors if we don't have a plane for them. */
-	if (!output->cursor_handle)
-		return 0;
-
-	for (i = 0; i < ARRAY_LENGTH(output->gbm_cursor_fb); i++) {
-		struct gbm_bo *bo;
-
-		if (gbm_device_get_fd(b->gbm) != output->device->kms_device->fd) {
-			output->gbm_cursor_fb[i] =
-				drm_fb_create_dumb(output->device,
-						   device->cursor_width,
-						   device->cursor_height,
-						   DRM_FORMAT_ARGB8888);
-			/* Override buffer type, since we know it is a cursor */
-			output->gbm_cursor_fb[i]->type = BUFFER_CURSOR;
-			output->gbm_cursor_handle[i] =
-				output->gbm_cursor_fb[i]->handles[0];
-		} else {
-			bo = gbm_bo_create(b->gbm, device->cursor_width, device->cursor_height,
-					   GBM_FORMAT_ARGB8888,
-					   GBM_BO_USE_CURSOR | GBM_BO_USE_WRITE);
-			if (!bo)
-				goto err;
-
-			output->gbm_cursor_fb[i] =
-				drm_fb_get_from_bo(bo, device, false, BUFFER_CURSOR);
-			if (!output->gbm_cursor_fb[i]) {
-				gbm_bo_destroy(bo);
-				goto err;
-			}
-			output->gbm_cursor_handle[i] = gbm_bo_get_handle(bo).s32;
-		}
-	}
-
-	return 0;
-
-err:
-	weston_log("cursor buffers unavailable, using vulkan cursors\n");
-	device->cursors_are_broken = true;
-	drm_output_fini_cursor_vulkan(output);
-	return -1;
-}
-
 static void
 create_gbm_surface(struct gbm_device *gbm, struct drm_output *output)
 {
@@ -542,8 +418,6 @@ drm_output_init_egl(struct drm_output *output, struct drm_backend *b)
 		return -1;
 	}
 
-	drm_output_init_cursor_egl(output, b);
-
 	return 0;
 }
 
@@ -744,8 +618,6 @@ drm_output_init_vulkan(struct drm_output *output, struct drm_backend *b)
 		return -1;
 	}
 
-	drm_output_init_cursor_vulkan(output, b);
-
 	return 0;
 }
 
@@ -766,7 +638,6 @@ drm_output_fini_egl(struct drm_output *output)
 	renderer->gl->output_destroy(&output->base);
 	gbm_surface_destroy(output->gbm_surface);
 	output->gbm_surface = NULL;
-	drm_output_fini_cursor_egl(output);
 }
 
 void
@@ -785,8 +656,6 @@ drm_output_fini_vulkan(struct drm_output *output)
 		renderer->destroy_renderbuffer(output->renderbuffer[i]);
 
 	renderer->vulkan->output_destroy(&output->base);
-
-	drm_output_fini_cursor_vulkan(output);
 }
 
 struct drm_fb *
diff --git a/libweston/backend-drm/drm-internal.h b/libweston/backend-drm/drm-internal.h
index 4f3c1c9bb..9140ddb03 100644
--- a/libweston/backend-drm/drm-internal.h
+++ b/libweston/backend-drm/drm-internal.h
@@ -645,8 +645,7 @@ struct drm_output {
 	/* True, if underlay planes exist. */
 	bool has_underlay;
 
-	uint32_t gbm_cursor_handle[2];
-	struct drm_fb *gbm_cursor_fb[2];
+	struct drm_fb *cursor_fb[2];
 	struct drm_plane_handle *cursor_handle;
 	int current_cursor;
 
@@ -945,7 +944,6 @@ drm_output_ensure_hdr_output_metadata_blob(struct drm_output *output);
 enum wdrm_colorspace
 wdrm_colorspace_from_output(struct weston_output *output);
 
-#ifdef BUILD_DRM_GBM
 extern struct drm_fb *
 drm_fb_get_from_paint_node(struct drm_output_state *state,
 			   struct weston_paint_node *pnode,
@@ -960,21 +958,6 @@ drm_fb_get_from_dmabuf_attributes(struct dmabuf_attributes *attributes,
 				  struct drm_device *device, bool is_opaque,
 				  bool direct_display, bool is_internal,
 				  uint32_t *try_view_on_plane_failure_reasons);
-#else
-static inline struct drm_fb *
-drm_fb_get_from_paint_node(struct drm_output_state *state,
-			   struct weston_paint_node *pnode,
-			   uint32_t *try_view_on_plane_failure_reasons)
-{
-	return NULL;
-}
-static inline bool
-drm_can_scanout_dmabuf(struct weston_backend *backend,
-		       struct linux_dmabuf_buffer *dmabuf)
-{
-	return false;
-}
-#endif
 
 struct drm_pending_state *
 drm_pending_state_alloc(struct drm_device *device);
diff --git a/libweston/backend-drm/drm.c b/libweston/backend-drm/drm.c
index 8e88eff5e..2a0ba88f8 100644
--- a/libweston/backend-drm/drm.c
+++ b/libweston/backend-drm/drm.c
@@ -867,7 +867,6 @@ err:
 	free(msg);
 }
 
-#ifdef BUILD_DRM_GBM
 /**
  * Update the image for the current cursor surface
  *
@@ -875,10 +874,9 @@ err:
  * @param ev Source view for cursor
  */
 static void
-cursor_bo_update(struct drm_output *output, struct weston_paint_node *pnode)
+cursor_update(struct drm_output *output, struct weston_paint_node *pnode)
 {
 	struct drm_device *device = output->device;
-	struct gbm_bo *bo = output->gbm_cursor_fb[output->current_cursor]->bo;
 	struct weston_buffer *buffer = pnode->surface->buffer_ref.buffer;
 	uint32_t buf[device->cursor_width * device->cursor_height];
 	uint8_t *s;
@@ -899,20 +897,8 @@ cursor_bo_update(struct drm_output *output, struct weston_paint_node *pnode)
 		       buffer->width * 4);
 	wl_shm_buffer_end_access(buffer->shm_buffer);
 
-	if (bo) {
-		if (gbm_bo_write(bo, buf, sizeof buf) < 0)
-			weston_log("failed update cursor: %s\n", strerror(errno));
-	} else {
-		memcpy(output->gbm_cursor_fb[output->current_cursor]->map,
-		       buf, sizeof buf);
-	}
+	memcpy(output->cursor_fb[output->current_cursor]->map, buf, sizeof buf);
 }
-#else
-static void
-cursor_bo_update(struct drm_output *output, struct weston_paint_node *pnode)
-{
-}
-#endif
 
 static void
 drm_output_prepare_repaint(struct weston_output *output_base)
@@ -973,13 +959,13 @@ drm_output_repaint(struct weston_output *output_base)
 			output->current_cursor++;
 			output->current_cursor =
 				output->current_cursor %
-					ARRAY_LENGTH(output->gbm_cursor_fb);
-			cursor_bo_update(output, cursor_node);
+					ARRAY_LENGTH(output->cursor_fb);
+			cursor_update(output, cursor_node);
 		}
 		pixman_region32_fini(&damage);
 
 		cursor_state->fb =
-			drm_fb_ref(output->gbm_cursor_fb[output->current_cursor]);
+			drm_fb_ref(output->cursor_fb[output->current_cursor]);
 		drm_fb_unref(old_fb);
 	}
 
@@ -2943,6 +2929,51 @@ should_wait_drm_events(struct drm_device *device)
 	return false;
 }
 
+static void drm_output_fini_cursor(struct drm_output *output)
+{
+	unsigned int i;
+
+	for (i = 0; i < ARRAY_LENGTH(output->cursor_fb); i++) {
+		if (!output->cursor_fb[i])
+			continue;
+
+		output->cursor_fb[i]->type = BUFFER_PIXMAN_DUMB;
+		drm_fb_unref(output->cursor_fb[i]);
+		output->cursor_fb[i] = NULL;
+	}
+}
+
+static int
+drm_output_init_cursor(struct drm_output *output, struct drm_backend *b)
+{
+	struct drm_device *device = output->device;
+	unsigned int i;
+
+	/* No point creating cursors if we don't have a plane for them. */
+	if (!output->cursor_handle)
+		return 0;
+
+	for (i = 0; i < ARRAY_LENGTH(output->cursor_fb); i++) {
+		output->cursor_fb[i] = drm_fb_create_dumb(output->device,
+							  device->cursor_width,
+							  device->cursor_height,
+							  DRM_FORMAT_ARGB8888);
+		if (!output->cursor_fb[i])
+			goto err;
+
+		/* Override buffer type, since we know it is a cursor */
+		output->cursor_fb[i]->type = BUFFER_CURSOR;
+	}
+
+	return 0;
+
+err:
+	weston_log("cursor buffers unavailable, using sw cursors\n");
+	device->cursors_are_broken = true;
+	drm_output_fini_cursor(output);
+	return -1;
+}
+
 static int
 drm_output_enable(struct weston_output *base)
 {
@@ -3007,6 +3038,8 @@ drm_output_enable(struct weston_output *base)
 		goto err_planes;
 	}
 
+	drm_output_init_cursor(output, b);
+
 	drm_output_init_backlight(output);
 
 	output->base.start_repaint_loop = drm_output_start_repaint_loop;
@@ -3065,6 +3098,8 @@ drm_output_deinit(struct weston_output *base)
 	else
 		drm_output_fini_egl(output);
 
+	drm_output_fini_cursor(output);
+
 	drm_output_detach_crtc(output);
 
 	output->blend_to_output_xform = NULL;
diff --git a/libweston/backend-drm/fb.c b/libweston/backend-drm/fb.c
index c5c4ee956..6fe2c4963 100644
--- a/libweston/backend-drm/fb.c
+++ b/libweston/backend-drm/fb.c
@@ -373,23 +373,27 @@ drm_fb_destroy_gbm(struct gbm_bo *bo, void *data)
 	       fb->type == BUFFER_CURSOR);
 	drm_fb_destroy(fb);
 }
+#endif
 
 static void
 drm_fb_destroy_dmabuf(struct drm_fb *fb)
 {
 	int i;
 
-	if (fb->bo) {
-		/* We deliberately do not close the GEM handles here; GBM manages
-		 * their lifetime through the BO. */
-		gbm_bo_destroy(fb->bo);
-	} else {
+	if (!fb->bo) {
 		for (i = 0; i < fb->num_planes; i++) {
 			struct drm_gem_close arg = { fb->handles[i], };
 			drmIoctl(fb->fd, DRM_IOCTL_GEM_CLOSE, &arg);
 		}
+		goto out;
 	}
 
+#ifdef BUILD_DRM_GBM
+	/* We deliberately do not close the GEM handles here; GBM manages
+	 * their lifetime through the BO. */
+	if (fb->bo)
+		gbm_bo_destroy(fb->bo);
+
 	/*
 	 * If we imported the dmabuf into a scanout device, we are responsible
 	 * for closing the GEM handle.
@@ -401,7 +405,9 @@ drm_fb_destroy_dmabuf(struct drm_fb *fb)
 		}
 
 	}
+#endif
 
+out:
 	for (i = 0; i < fb->num_duped_fds; i++)
 		close(fb->fds[i]);
 
@@ -414,16 +420,11 @@ drm_fb_get_from_dmabuf_attributes(struct dmabuf_attributes *attributes,
 				  bool direct_display, bool is_internal,
 				  uint32_t *try_view_on_plane_failure_reasons)
 {
+#ifdef BUILD_DRM_GBM
 	struct drm_backend *backend = device->backend;
+#endif
 	struct drm_fb *fb;
 	int i;
-	struct gbm_import_fd_modifier_data import_mod = {
-		.width = attributes->width,
-		.height = attributes->height,
-		.format = attributes->format,
-		.num_fds = attributes->n_planes,
-		.modifier = attributes->modifier,
-	};
 
 #if 0
 	/* We should not import to KMS a buffer that has been allocated using no
@@ -461,10 +462,6 @@ drm_fb_get_from_dmabuf_attributes(struct dmabuf_attributes *attributes,
 	fb->type = is_internal ? BUFFER_DMABUF_BACKEND : BUFFER_DMABUF;
 	fb->backend = device->backend;
 
-	ARRAY_COPY(import_mod.fds, attributes->fd);
-	ARRAY_COPY(import_mod.strides, attributes->stride);
-	ARRAY_COPY(import_mod.offsets, attributes->offset);
-
 	/* skip bo import if dmabuf is using direct-display extension */
 	if (direct_display) {
 		fb->direct_display = true;
@@ -479,14 +476,8 @@ drm_fb_get_from_dmabuf_attributes(struct dmabuf_attributes *attributes,
 			}
 			fb->num_duped_fds++;
 		}
-
-		goto bo_import_skip;
 	}
 
-	fb->bo = gbm_bo_import(backend->gbm, GBM_BO_IMPORT_FD_MODIFIER,
-			       &import_mod, GBM_BO_USE_SCANOUT);
-
-bo_import_skip:
 	fb->width = attributes->width;
 	fb->height = attributes->height;
 	fb->modifier = attributes->modifier;
@@ -515,6 +506,25 @@ bo_import_skip:
 	}
 
 	fb->num_planes = attributes->n_planes;
+
+#ifdef BUILD_DRM_GBM
+	if (backend->gbm && !direct_display) {
+		struct gbm_import_fd_modifier_data import_mod = {
+			.width = attributes->width,
+			.height = attributes->height,
+			.format = attributes->format,
+			.num_fds = attributes->n_planes,
+			.modifier = attributes->modifier,
+		};
+
+		ARRAY_COPY(import_mod.fds, attributes->fd);
+		ARRAY_COPY(import_mod.strides, attributes->stride);
+		ARRAY_COPY(import_mod.offsets, attributes->offset);
+
+		fb->bo = gbm_bo_import(backend->gbm, GBM_BO_IMPORT_FD_MODIFIER,
+				       &import_mod, GBM_BO_USE_SCANOUT);
+	}
+
 	if (fb->bo) {
 		for (i = 0; i < attributes->n_planes; i++) {
 			union gbm_bo_handle handle;
@@ -527,9 +537,12 @@ bo_import_skip:
 			}
 			fb->handles[i] = handle.u32;
 		}
-	} else {
+	}
+#endif
+
+	if (!fb->bo) {
 		for (i = 0; i < fb->num_planes; i++) {
-			if (drmPrimeFDToHandle(fb->fd, import_mod.fds[i],
+			if (drmPrimeFDToHandle(fb->fd, attributes->fd[i],
 					       &fb->handles[i])) {
 				*try_view_on_plane_failure_reasons |=
 					FAILURE_REASONS_GBM_BO_IMPORT_FAILED;
@@ -564,6 +577,7 @@ drm_fb_get_from_dmabuf(struct linux_dmabuf_buffer *dmabuf,
 						 try_view_on_plane_failure_reasons);
 }
 
+#ifdef BUILD_DRM_GBM
 struct drm_fb *
 drm_fb_get_from_bo(struct gbm_bo *bo, struct drm_device *device,
 		   bool is_opaque, enum drm_fb_type type)
@@ -649,26 +663,25 @@ drm_fb_unref(struct drm_fb *fb)
 	case BUFFER_PIXMAN_DUMB:
 		drm_fb_destroy_dumb(fb);
 		break;
-#ifdef BUILD_DRM_GBM
 	case BUFFER_CURSOR:
+#ifdef BUILD_DRM_GBM
 	case BUFFER_CLIENT:
 		gbm_bo_destroy(fb->bo);
 		break;
 	case BUFFER_GBM_SURFACE:
 		gbm_surface_release_buffer(fb->gbm_surface, fb->bo);
 		break;
+#endif
 	case BUFFER_DMABUF:
 	case BUFFER_DMABUF_BACKEND:
 		drm_fb_destroy_dmabuf(fb);
 		break;
-#endif
 	default:
 		assert(NULL);
 		break;
 	}
 }
 
-#ifdef BUILD_DRM_GBM
 bool
 drm_can_scanout_dmabuf(struct weston_backend *backend,
 		       struct linux_dmabuf_buffer *dmabuf)
@@ -799,17 +812,16 @@ drm_fb_get_from_paint_node(struct drm_output_state *state,
 	buf_fb->device = device;
 	wl_list_insert(&private->buffer_fb_list, &buf_fb->link);
 
-	/* GBM is used for dmabuf import as well as from client wl_buffer. */
-	if (!b->gbm) {
-		*try_view_on_plane_failure_reasons |= FAILURE_REASONS_NO_GBM;
-		goto unsuitable;
-	}
-
 	if (buffer->type == WESTON_BUFFER_DMABUF) {
 		fb = drm_fb_get_from_dmabuf(buffer->dmabuf, device, is_opaque,
 					    &buf_fb->failure_reasons);
 		if (!fb)
 			goto unsuitable;
+#ifdef BUILD_DRM_GBM
+	} else if (!b->gbm) {
+		/* GBM is used for dmabuf import as well as from client wl_buffer. */
+		*try_view_on_plane_failure_reasons |= FAILURE_REASONS_NO_GBM;
+		goto unsuitable;
 	} else if (buffer->type == WESTON_BUFFER_RENDERER_OPAQUE) {
 		struct gbm_bo *bo;
 
@@ -825,6 +837,7 @@ drm_fb_get_from_paint_node(struct drm_output_state *state,
 			gbm_bo_destroy(bo);
 			goto unsuitable;
 		}
+#endif
 	} else {
 		*try_view_on_plane_failure_reasons |= FAILURE_REASONS_BUFFER_TYPE;
 		goto unsuitable;
@@ -858,4 +871,3 @@ unsuitable:
 	*try_view_on_plane_failure_reasons |= buf_fb->failure_reasons;
 	return NULL;
 }
-#endif
diff --git a/libweston/backend-drm/kms.c b/libweston/backend-drm/kms.c
index e773ee304..5ef6f81ca 100644
--- a/libweston/backend-drm/kms.c
+++ b/libweston/backend-drm/kms.c
@@ -872,11 +872,11 @@ drm_output_set_cursor(struct drm_output_state *output_state)
 		return;
 	}
 
-	assert(state->fb == output->gbm_cursor_fb[output->current_cursor]);
+	assert(state->fb == output->cursor_fb[output->current_cursor]);
 	assert(!plane->state_cur->handle ||
 	       plane->state_cur->handle->output == output);
 
-	handle = output->gbm_cursor_handle[output->current_cursor];
+	handle = output->cursor_fb[output->current_cursor]->handles[0];
 	if (plane->state_cur->fb != state->fb) {
 		if (drmModeSetCursor(device->kms_device->fd, crtc->crtc_id, handle,
 				     device->cursor_width, device->cursor_height)) {
diff --git a/libweston/backend-drm/state-propose.c b/libweston/backend-drm/state-propose.c
index bb72f4c15..dc1fd0800 100644
--- a/libweston/backend-drm/state-propose.c
+++ b/libweston/backend-drm/state-propose.c
@@ -225,7 +225,6 @@ out:
 	return NULL;
 }
 
-#ifdef BUILD_DRM_GBM
 static struct drm_plane_state *
 drm_output_prepare_cursor_paint_node(struct drm_output_state *output_state,
 				     struct weston_paint_node *pnode,
@@ -252,9 +251,6 @@ drm_output_prepare_cursor_paint_node(struct drm_output_state *output_state,
 
 	p_name = drm_output_get_handle_type_name(handle);
 
-	/* We use GBM to import SHM buffers. */
-	assert(b->gbm);
-
 	plane_state = drm_output_state_get_plane(output_state, plane);
 	assert(!plane_state->fb);
 
@@ -320,7 +316,7 @@ out:
 	 * Later when we determine if the cursor needs an update, we'll
 	 * select the correct fb to use.
 	 */
-	plane_state->fb = drm_fb_ref(output->gbm_cursor_fb[0]);
+	plane_state->fb = drm_fb_ref(output->cursor_fb[0]);
 
 	drm_debug(b, "\t\t\t\t[%s] provisionally assigned paint node %s to cursor\n",
 		  p_name, pnode->internal_name);
@@ -334,15 +330,6 @@ err:
 	drm_plane_state_put_back(plane_state);
 	return NULL;
 }
-#else
-static struct drm_plane_state *
-drm_output_prepare_cursor_paint_node(struct drm_output_state *output_state,
-				     struct weston_paint_node *node,
-				     uint64_t zpos)
-{
-	return NULL;
-}
-#endif
 
 static void
 drm_output_check_zpos_plane_states(struct drm_output_state *state)
@@ -1388,10 +1375,6 @@ drm_output_propose_state(struct weston_output *output_base,
 			  pnode->internal_name, output->base.name,
 			  (unsigned long) output->base.id);
 
-		if (!b->gbm)
-			pnode->try_view_on_plane_failure_reasons |=
-				FAILURE_REASONS_NO_GBM;
-
 		if (!weston_paint_node_has_valid_buffer(pnode))
 			pnode->try_view_on_plane_failure_reasons |=
 				FAILURE_REASONS_NO_BUFFER;
@@ -1589,7 +1572,7 @@ drm_assign_planes(struct weston_output *output_base)
 		drm_debug(b, "\t[repaint] could not reuse prior state\n");
 
 	if (!state && !device->disable_client_buffer_scanout &&
-	    !output_base->disable_planes && !output->is_virtual && b->gbm &&
+	    !output_base->disable_planes && !output->is_virtual &&
 	    !output_base->mirroring) {
 		drm_debug(b, "\t[repaint] trying planes-only build state\n");
 		mode = DRM_OUTPUT_PROPOSE_STATE_PLANES_ONLY;
-- 
2.39.5

